File: themeforest-Fj3VzBCp-fuse-angularjs-material-design-admin-template/fuse/src/app/views/dashboard/mi-proyecto/services/mi-proyecto.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '@/environments/environment';
import { END_POINTS } from '@/app/providers/utils/end-points';
import { ActividadItem, ObjetivoItem, PartidaItem } from '../models/proyecto.model';

@Injectable({ providedIn: 'root' })
export class MiProyectoService {
  private _http = inject(HttpClient);
  private readonly base = environment.url + END_POINTS.proyecto.miProyecto;

  editor$(): Observable<any> {
    return this._http.get<any>(this.base);
  }

  guardarCampo$(campo: string, valor: string): Observable<any> {
    return this._http.put<any>(`${this.base}/campos/${campo}`, { valor });
  }

  setEnfoque$(enfoque: string): Observable<any> {
    return this._http.put<any>(`${this.base}/enfoque`, { enfoque });
  }

  desbloquearEnfoque$(): Observable<any> {
    return this._http.post<any>(`${this.base}/enfoque/desbloquear`, {});
  }

  setFinanciamiento$(financiamiento: string): Observable<any> {
    return this._http.put<any>(`${this.base}/financiamiento`, { financiamiento });
  }

  // ── Objetivos ──
  agregarObjetivo$(o: ObjetivoItem): Observable<any> {
    return this._http.post<any>(`${this.base}/objetivos`, o);
  }
  actualizarObjetivo$(id: string, o: ObjetivoItem): Observable<any> {
    return this._http.put<any>(`${this.base}/objetivos/${id}`, o);
  }
  eliminarObjetivo$(id: string): Observable<any> {
    return this._http.delete<any>(`${this.base}/objetivos/${id}`);
  }

  // ── Actividades ──
  agregarActividad$(a: ActividadItem): Observable<any> {
    return this._http.post<any>(`${this.base}/actividades`, a);
  }
  actualizarActividad$(id: string, a: ActividadItem): Observable<any> {
    return this._http.put<any>(`${this.base}/actividades/${id}`, a);
  }
  eliminarActividad$(id: string): Observable<any> {
    return this._http.delete<any>(`${this.base}/actividades/${id}`);
  }

  // ── Presupuesto ──
  agregarPartida$(p: PartidaItem): Observable<any> {
    return this._http.post<any>(`${this.base}/partidas`, p);
  }
  actualizarPartida$(id: string, p: PartidaItem): Observable<any> {
    return this._http.put<any>(`${this.base}/partidas/${id}`, p);
  }
  eliminarPartida$(id: string): Observable<any> {
    return this._http.delete<any>(`${this.base}/partidas/${id}`);
  }

  // ── Hitos ──
  marcarListo$(): Observable<any> {
    return this._http.post<any>(`${this.base}/listo-revision`, {});
  }
  reenviarRevision$(): Observable<any> {
    return this._http.post<any>(`${this.base}/reenviar-revision`, {});
  }
  publicarPlan$(): Observable<any> {
    return this._http.post<any>(`${this.base}/plan/publicar`, {});
  }
  corregir$(campo: string, respuesta: string): Observable<any> {
    return this._http.post<any>(`${this.base}/observaciones/${campo}/corregir`, { respuesta });
  }
  responderRevisor$(revisorId: string, respuesta: string): Observable<any> {
    return this._http.post<any>(`${this.base}/revisores/${revisorId}/responder`, { respuesta });
  }
  solicitarJuradoInformante$(): Observable<any> {
    return this._http.post<any>(`${this.base}/jurado-informante/solicitar`, {});
  }
  responderJuradoInforme$(revisorId: string, respuesta: string): Observable<any> {
    return this._http.post<any>(`${this.base}/jurado-informante/${revisorId}/responder`, { respuesta });
  }
  subirTurnitin$(file: File, porcentaje: number): Observable<any> {
    const fd = new FormData();
    fd.append('archivo', file, file.name);
    fd.append('porcentaje', String(porcentaje));
    return this._http.post<any>(`${this.base}/turnitin`, fd);
  }
  subirProyectoFinal$(file: File): Observable<any> {
    const fd = new FormData();
    fd.append('archivo', file, file.name);
    return this._http.post<any>(`${this.base}/proyecto-final`, fd);
  }
  subirInformeFinal$(file: File): Observable<any> {
    const fd = new FormData();
    fd.append('archivo', file, file.name);
    return this._http.post<any>(`${this.base}/informe-final`, fd);
  }
  /** Ver/descargar un documento propio subido (turnitin | proyecto-final). */
  descargarDocumento$(tipo: 'turnitin' | 'proyecto-final'): Observable<Blob> {
    return this._http.get(`${this.base}/documentos/${tipo}`, { responseType: 'blob' });
  }
  solicitarAprobacion$(): Observable<any> {
    return this._http.post<any>(`${this.base}/expediente/solicitar-aprobacion`, {});
  }

  // ── Referencias ──
  /** Autocompletar una referencia desde su DOI (CrossRef). */
  buscarDoi$(doi: string): Observable<any> {
    return this._http.get<any>(`${this.base}/referencias/por-doi`, { params: { doi } });
  }
  /** Buscar candidatos de referencia por título/autor (sin DOI). */
  buscarTitulo$(q: string): Observable<any> {
    return this._http.get<any>(`${this.base}/referencias/buscar`, { params: { q } });
  }
  /** Rellenar una referencia pegando su BibTeX. */
  parsearBibtex$(bibtex: string): Observable<any> {
    return this._http.post<any>(`${this.base}/referencias/bibtex`, bibtex, { headers: { 'Content-Type': 'text/plain' } });
  }
  agregarReferencia$(r: any): Observable<any> {
    return this._http.post<any>(`${this.base}/referencias`, r);
  }
  actualizarReferencia$(id: string, r: any): Observable<any> {
    return this._http.put<any>(`${this.base}/referencias/${id}`, r);
  }
  eliminarReferencia$(id: string): Observable<any> {
    return this._http.delete<any>(`${this.base}/referencias/${id}`);
  }
  setEstiloCita$(estilo: string): Observable<any> {
    return this._http.put<any>(`${this.base}/estilo-cita/${estilo}`, {});
  }
  desbloquearEstiloCita$(): Observable<any> {
    return this._http.post<any>(`${this.base}/estilo-cita/desbloquear`, {});
  }
  /** Genera el documento del proyecto (formato = pdf | docx) y lo devuelve como Blob. */
  generarDocumento$(formato: 'pdf' | 'docx'): Observable<Blob> {
    return this._http.get(`${this.base}/documento/${formato}`, { responseType: 'blob' });
  }

  // ── DEMO / pruebas ──
  seedDemo$(): Observable<any> {
    return this._http.post<any>(`${this.base}/demo/seed`, {});
  }
  resetDemo$(): Observable<any> {
    return this._http.post<any>(`${this.base}/demo/reset`, {});
  }
}

File: lamb-upeu-sis/src/main/java/unmsm/edu/pe/tesis/domain/services/MiProyectoService.java
package unmsm.edu.pe.tesis.domain.services;

import unmsm.edu.pe.tesis.application.dto.*;

import java.util.UUID;

/** Casos de uso del ESTUDIANTE sobre su proyecto de tesis (Etapa 4). */
public interface MiProyectoService {

    ProyectoEditorResponse editor();

    /** La rúbrica (en blanco) con la que se evaluará su proyecto, según su enfoque. */
    unmsm.edu.pe.tesis.application.dto.RubricaAlumnoResponse rubrica();

    /** El Word oficial vigente de esa rúbrica. */
    ArchivoDescargable rubricaDocumento();

    void guardarCampo(String campo, String valor);

    void setEnfoque(String enfoque);

    /** Desbloquea el enfoque para poder cambiarlo. */
    void desbloquearEnfoque();

    void setFinanciamiento(String financiamiento);

    ObjetivoItem agregarObjetivo(ObjetivoRequest req);
    void actualizarObjetivo(UUID id, ObjetivoRequest req);
    void eliminarObjetivo(UUID id);

    ActividadItem agregarActividad(ActividadRequest req);
    void actualizarActividad(UUID id, ActividadRequest req);
    void eliminarActividad(UUID id);

    // ── Referencias bibliográficas (estructuradas, auto-formateadas) ──
    /** Consulta un DOI en CrossRef y devuelve los campos para autocompletar (no persiste). */
    ReferenciaRequest buscarReferenciaPorDoi(String doi);
    /** Busca candidatos por título/autor (sin DOI) para autocompletar. */
    java.util.List<ReferenciaRequest> buscarReferenciasPorTitulo(String q);
    /** Parsea una entrada BibTeX (Google Scholar / repositorios) a una referencia. */
    ReferenciaRequest parsearBibtex(String bibtex);
    ReferenciaItem agregarReferencia(ReferenciaRequest req);
    void actualizarReferencia(UUID id, ReferenciaRequest req);
    void eliminarReferencia(UUID id);
    /** Fija el estilo de cita del proyecto (queda bloqueado; un documento usa un solo estilo). */
    void setEstiloCita(String estilo);
    /** Desbloquea el estilo de cita para poder cambiarlo. */
    void desbloquearEstiloCita();
    /** Genera el documento del proyecto (formato = pdf | docx) con las referencias formateadas. */
    ArchivoDescargable exportarDocumento(String formato);

    PartidaItem agregarPartida(PartidaRequest req);
    void actualizarPartida(UUID id, PartidaRequest req);
    void eliminarPartida(UUID id);

    /** Marca el proyecto como listo para revisión (exige avance 100%). */
    void marcarListoRevision();

    /** Reenvía el proyecto a revisión tras corregir observaciones (de OBSERVADO a EN_REVISION). */
    void reenviarRevision();

    /** Publica el plan de actividades (exige al menos una actividad). */
    void publicarPlan();

    /** Sube el informe de similitud de Turnitin (carga manual) + porcentaje. */
    void subirTurnitin(byte[] contenido, String nombreOriginal, String contentType, Integer porcentaje);

    /** Sube el proyecto en versión final (PDF) para el expediente. */
    void subirProyectoFinal(byte[] contenido, String nombreOriginal, String contentType);

    /** Sube el informe final de la tesis (Etapa 6 · ejecución). */
    void subirInformeFinal(byte[] contenido, String nombreOriginal, String contentType);

    /** Descarga un documento propio que el estudiante subió (tipo = turnitin | proyecto-final). */
    ArchivoDescargable descargarDocumento(String tipo);

    /** El estudiante confirma la corrección de un ítem observado. */
    void corregirItem(String campo, String respuesta);

    /** El estudiante responde (levanta) las observaciones de un revisor del proyecto (Etapa 5). */
    void responderRevisor(UUID revisorId, String respuesta);

    /** El estudiante solicita el Jurado Informante del informe final (Etapa 7). */
    void solicitarJuradoInformante();

    /** El estudiante responde (levanta) las observaciones de un miembro del Jurado Informante (Etapa 7). */
    void responderJuradoInforme(UUID revisorId, String respuesta);

    /** Sube el expediente y genera la solicitud de aprobación (fin de la Etapa 4). */
    void solicitarAprobacion();

    /** [DEMO] Rellena el proyecto con datos de prueba y lo envía a revisión (para probar el flujo rápido). */
    void seedDemo();

    /** [DEMO] Reinicia el flujo: borra observaciones/correcciones y deja el proyecto como "aún no enviado". */
    void resetDemo();
}
